File: src/main/java/net/lifecity/mc/skillmaster/skill/SeparatedSkill.java
package net.lifecity.mc.skillmaster.skill;

import lombok.Getter;
import net.lifecity.mc.skillmaster.SkillMaster;
import net.lifecity.mc.skillmaster.user.SkillUser;
import net.lifecity.mc.skillmaster.weapon.Weapon;
import org.bukkit.ChatColor;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.List;

/**
 * 複数の入力ができる複合スキルのスーパークラス
 */
public abstract class SeparatedSkill extends Skill {

    protected final int activationTime;

    @Getter
    protected boolean activated = false;

    protected SeparatedSkill(String name, Weapon weapon, SkillType type, List<String> lore, int num, int point, int activationTime, int interval, SkillUser user) {
        super(name, weapon, type, lore, num, point, interval, user);
        this.activationTime = activationTime;
    }

    /**
     * スキルを発動します
     */
    public void activate() {
        // ログ
        user.sendActionBar(ChatColor.DARK_AQUA + "複合スキル『" + name + "』発動");

        // 発動中にする
        activated = true;

        // 終了処理
        new ActivationTimer();
    }

    /**
     * 発動している期間を計るクラス
     */
    private class ActivationTimer extends BukkitRunnable {

        private int count = 0;

        public ActivationTimer() {
            runTaskTimer(SkillMaster.instance, 0, 1);
        }

        @Override
        public void run() {
            if (!activated) //発動中か確認
                cancel();

            if (count >= activationTime) { //カウント確認
                deactivate();
                cancel();
            }

            count++;
        }
    }

    /**
     * 追加入力の処理
     */
    public abstract void additionalInput();

    /**
     * スキルを終了します
     */
    public void deactivate() {
        if (!activated) //発動していなかったら戻る
            return;

        activated = false; //非発動化する

        // ログ
        user.sendActionBar(ChatColor.RED + "複合スキル『" + name + "』終了");

        //インターバル処理
        super.deactivate();
    }
}

File: src/main/java/net/lifecity/mc/skillmaster/skill/SeparatedSkill.kt
package net.lifecity.mc.skillmaster.skill

import com.github.syari.spigot.api.scheduler.runTaskTimer
import net.lifecity.mc.skillmaster.SkillMaster
import net.lifecity.mc.skillmaster.user.SkillUser
import net.lifecity.mc.skillmaster.weapon.Weapon
import org.bukkit.ChatColor
import org.bukkit.scheduler.BukkitRunnable

abstract class SeparatedSkill(
    name: String,
    weaponList: List<Weapon>,
    type: SkillType,
    lore: List<String>,
    val activationTime: Int,
    interval: Int,
    user: SkillUser?,
    val canCancel: Boolean = true
) : Skill(name, weaponList, type, lore, interval, user = user) {

    var activated = false

    override fun activate() {
        // ログ
        user?.sendActionBar("" + ChatColor.DARK_AQUA + "複合スキル『" + name + "』発動")

        // 発動中にする
        activated = true

        //終了処理：ActivationTimer起動
        var count = 0
        SkillMaster.INSTANCE.runTaskTimer(1) {
            if (!this@SeparatedSkill.activated) //発動中か確認
                cancel()

            if (count >= activationTime) { //カウント確認
                deactivate()
                cancel()
            }

            count++
        }
    }

    /**
     * 追加入力の処理
     */
    abstract fun additionalInput()

    /**
     * スキルを終了します
     */
    override fun deactivate() {
        if (!activated) //発動していなかったら戻る
            return

        activated = false //発動解除

        // ログ
        user?.sendActionBar("" + ChatColor.RED + "複合スキル『" + name + "』終了")

        // インターバル処理
        super.deactivate()
    }

    override fun init() {
        super.init()

        activated = false
    }
}